hircluster.c

```diff
@@ -3234,15 +3234,73 @@ int redisClusterAppendFormattedCommand(redisClusterContext *cc, char *cmd,
     slot_num = command_format_by_slot(cc, command, commands);
 
     if (slot_num < 0) {
-        goto error;
+        if (command->type == CMD_REQ_REDIS_MULTI) {
+            // This cmd will not be sent until we get a slot from next cmd
+            if (cc->transaction_slot != TRANSACTION_NONE) {
+                __redisClusterSetError(cc, REDIS_ERR_OTHER,
+                                       "transaction already started");
+                goto error;
+            }
+            cc->transaction_slot = TRANSACTION_STARTED;
+            command->cmd = strdup(cmd); // Keep copy of the command to send
+            goto done;
+        } else if (command->type == CMD_REQ_REDIS_EXEC ||
+                   command->type == CMD_REQ_REDIS_DISCARD) {
+            if (cc->transaction_slot == TRANSACTION_NONE) {
+                __redisClusterSetError(cc, REDIS_ERR_OTHER,
+                                       "transaction not started");
+                goto error;
+            }
+            command->slot_num = cc->transaction_slot;
+            cc->transaction_slot = TRANSACTION_NONE;
+        } else {
+            __redisClusterSetError(
+                cc, REDIS_ERR_OTHER,
+                "No keys in command(must have keys for redis cluster mode)");
+            goto error;
+        }
     } else if (slot_num >= REDIS_CLUSTER_SLOTS) {
         __redisClusterSetError(cc, REDIS_ERR_OTHER, "slot_num is out of range");
         goto error;
     }
 
+    if (cc->transaction_slot >= 0 && cc->transaction_slot != slot_num) {
+        __redisClusterSetError(cc, REDIS_ERR_OTHER,
+                               "only supports same slot transactions");
+        goto error;
+    }
+
+    if (cc->transaction_slot == TRANSACTION_STARTED) {
+        // Previous command was MULTI, use current slot as transaction slot
+        listNode *prev_command = listLast(cc->requests);
+        if (prev_command == NULL) {
+            __redisClusterSetError(cc, REDIS_ERR_OTHER, "transaction failure");
+            goto error;
```

Comment:
Can prev_command ever be NULL if we are in TRANSACTION_STARTED? (Maybe after reconnect?)

If it can't happen, maybe this should be an ASSERT instead?

Reply:
Should not be possible unless there is some fishyness. There are a lot of checks throughout the codebase and many where asserts could be more fitting, like if the clustercontext is null..which is everywhere.
I was thinking that when there are more tests to lean on we could make the internals more strict and assert where applicable